xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -128,17 +136,775 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
+	logicalDNSClusterName := "logical_dns_cluster"
+	clusterName := "aggregate_cluster"
+
+	backends, stop := startTestServiceBackends(t, 1)
+	defer stop()
+
+	nonExistingBackend1, stop1 := startTestServiceBackends(t, 2)
+	nonExistingBackend2, stop2 := startTestServiceBackends(t, 2)
+	stop1()
+	stop2()
+
+	endpoints := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: edsClusterName,
+		Localities: []e2e.LocalityOptions{{
+			Name:     "locality0",
+			Weight:   1,
+			Backends: backendOptions(t, nonExistingBackend1),
+			Priority: 0,
+		}, {
+			Name:     "locality1",
+			Weight:   1,
+			Backends: backendOptions(t, nonExistingBackend2),
+			Priority: 1,
+		}},
+	})
+	edsCluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: edsClusterName,
+		ServiceName: edsClusterName,
+	})
+
+	logicalDNSCluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		Type:        e2e.ClusterTypeLogicalDNS,
+		ClusterName: logicalDNSClusterName,
+		DNSHostName: "server.example.com",
+		DNSPort:     443,
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		Type:        e2e.ClusterTypeAggregate,
+		// TODO: when "A75: xDS Aggregate Cluster Behavior Fixes" is merged, the
+		// policy will have to be set on the child clusters.
+		Policy:     e2e.LoadBalancingPolicyRingHash,
+		ChildNames: []string{edsClusterName, logicalDNSClusterName},
+	})
+	route := channelIDHashRoute("new_route", virtualHostName, clusterName)
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	err := xdsServer.Update(context.Background(), e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{endpoints},
+		Clusters:  []*v3clusterpb.Cluster{cluster, edsCluster, logicalDNSCluster},
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	_, _, _, dnsR := e2e.SetupDNS(t)
+	dnsR.InitialState(resolver.State{Addresses: []resolver.Address{{Addr: backends[0].Address}}})
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// TODO: c-core injects a 500ms RPC delay here: https://github.com/grpc/grpc/blob/master/test/cpp/end2end/xds/xds_ring_hash_end2end_test.cc#L220
+	// Should we try to do the same? do we have a way to do that in Go? Perhaps using something like https://github.com/grpc/grpc-go/blob/04ea82009cdb9ecdefc6289f4c93ec919a10b3b6/benchmark/latency/latency.go#L76-L75?
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	if _, err = client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Errorf("rpc UnaryCall() failed: %v", err)
+	}
```

Comment:
Actually, this is the same comment as you gave me for the first test -- c-core doesn't make this check, presumably because there is a single listening backend, so there's no way RPC succeed if they go to the wrong backend. I added the check though, I think it makes sense to have it.